### PR TITLE
Deprecate ' ', '/', '\' and ':' in project and other domain object names

### PR DESCRIPTION
**Pull request by jjohannes**
Discussion: https://github.com/gradle/gradle-private/issues/167

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    private static final char[] FORBIDDEN_CHARACTERS = new char[] {' ', '/', '\\', ':'};
+    private static final char REPLACEMENT_CHARACTER = '_';
+
+    private NameValidator() { }
+
+    /**
+     * Validates that a given name string does not contain any forbidden characters.
+     */
+    public static void validate(String name) {
+        if (StringUtils.containsNone(name, FORBIDDEN_CHARACTERS)) {
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: :x: This logic might read better as

```
if (StringUtils.containsAny(name, FORBIDDEN_CHARACTERS)) {
     DeprecationLogger. nagUserOfDeprecated(...)
}
```

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import org.gradle.test.fixtures.file.TestDirectoryProvider
+import org.gradle.test.fixtures.file.TestFile
+
+import static org.gradle.internal.os.OperatingSystem.WINDOWS
+import static org.gradle.internal.os.OperatingSystem.current
+
+class NameValidationIntegrationTest extends AbstractIntegrationSpec {
+
+    def setup() {
+        executer.expectDeprecationWarning()
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: :x: We should assert the deprecation warning output for all test cases.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.util;
+
+import org.apache.commons.lang.StringUtils;
+
+import java.util.Arrays;
+
+public final class NameValidator {
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: Might make sense to write a unit test for this class to cover the basis. The integration tests do not test all forbidden characters for all test cases.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    private NameValidator() { }
+
+    /**
+     * Validates that a given name string does not contain any forbidden characters.
+     */
+    public static void validate(String name) {
+        if (StringUtils.containsNone(name, FORBIDDEN_CHARACTERS)) {
+            return;
+        }
+        DeprecationLogger.nagUserOfDeprecated("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS) + ". This");
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: :x: I believe you have some stray word here after the sentence: " This".

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java [hunk 5]
```diff
@@ -27,6 +27,7 @@
     private static final String NON_PRECOMPOSED_NON_ASCII = "-√æ∫ʙ₦∆√∫";
     private static final String FILESYSTEM_RESERVED_CHARS = "-./\\?%*:|\"<>";
     private static final String XML_MARKUP_CHARS = "-<with>some<xml-markup/></with>";
+    private static final String GRADLE_NAME_FORBIDDEN_CHARACTERS = " /\\:"; //See: NameValidator.FORBIDDEN_CHARACTERS
```
Comment (bmuschko) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java]: :o: Maybe use Javadocs instead with the `@see` annotation.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        succeeds 'help'
+    }
+
+    def "subproject names should not contain forbidden characters"() {
+        when:
+        settingsFile << "include 'folder:name with spaces'"
+
+        then:
+        succeeds 'help'
+        output.contains("The name 'name with spaces' contains at least one of the following characters:")
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: :x: We should verify the full deprecation sentence.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        buildFile << "configurations { 'some/really.\\\\strange name:' {} }"
+
+        then:
+        succeeds 'help'
+    }
+
+    def "does not assign an invalid project name from folder name"() {
+        given:
+        def buildFolder = file(current() == WINDOWS ? "folder  name" : "folder: name")
+        GradleExecuter subDirExecuter = new GradleContextualExecuter(distribution, new TestDirectoryProvider() {
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: :x: Instead of creating a new `GradleExecuter` instance couldn't we just point it to the sub directory via `inDirectory(buildFolder)` or `executer.inDirectory(buildFolder)` ?

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            TestFile getTestDirectory() { return buildFolder }
+            void suppressCleanup() { }
+        }, getBuildContext())
+
+        when:
+        buildFolder.file("build.gradle") << "println rootProject.name"
+        result = subDirExecuter.withTasks('help').run()
+
+        then:
+        output.contains("folder__name")
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: :o: I might be wrong here but couldn't actually check the task path via `executedTasks(':folder__name:help')`?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java [hunk 9]
```diff
@@ -42,7 +42,7 @@
 import static org.junit.Assert.assertThat;
 
 public class OutputScrapingExecutionResult implements ExecutionResult {
-    static final Pattern STACK_TRACE_ELEMENT = Pattern.compile("\\s+(at\\s+)?([\\w.$_]+/)?[\\w.$_]+\\.[\\w$_ =\\+\'-]+\\(.+?\\)");
+    static final Pattern STACK_TRACE_ELEMENT = Pattern.compile("\\s+(at\\s+)?([\\w.$_]+/)?[\\w.$_]+\\.[\\w$_ =\\+\'-<>]+\\(.+?\\)");
```
Comment (bmuschko) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java]: Can you explain why this change is needed?

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        when:
+        buildFile << "configurations { 'some/really.\\\\strange name:' {} }"
+
+        then:
+        succeeds 'help'
+    }
+
+    def "does not assign an invalid project name from folder name"() {
+        given:
+        def buildFolder = file(current() == WINDOWS ? "folder  name" : "folder: name")
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: Why do we need special handling for different OSes?

**Comment by bmuschko** [comment]
:x: We should also mention this in the release notes as potentially breaking change.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java [hunk 11]
```diff
@@ -64,6 +65,7 @@ public T create(String name, Closure configureClosure) {
 
     public T create(String name, Action<? super T> configureAction) throws InvalidUserDataException {
         assertCanAdd(name);
+        NameValidator.validate(name);
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java]: This is a pretty deep change, that will impact a lot of things, including custom plugins. I don't think we can really claim to know what constitutes a valid name in every domain, and we should perhaps apply this validation only to those thing we know about.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java [hunk 12]
```diff
@@ -154,6 +155,7 @@ private AbstractTask(TaskInfo taskInfo) {
         this.publicType = taskInfo.publicType;
         assert project != null;
         assert name != null;
+        NameValidator.validate(name);
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java]: I think it would be better to do this in the task factory, where we do some other validation.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 10 above)
Comment (jjohannes) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: `:` is not allowed in file names on Windows. But I thought it would be good to have a coverage for such a file name that only works on Linux here.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 7 above)
Comment (jjohannes) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: Cool. I did not find the `inDirectory()` method. Probably because I was expecting it to be called `with...`. Or I am just blind... 
Thanks for the hint! This makes the test code much simpler.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java (hunk 4 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: No. More will be appended by the _nag_ method. I'll extend the integration tests, as you suggest below, to clarify and test that this actually is the case.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java (hunk 1 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: Agreed.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 2 above)
Comment (jjohannes) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: Agreed.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 6 above)
Comment (jjohannes) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: Yes that's better. I added that to all tests.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 8 above)
Comment (jjohannes) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: No the root project task is printed `:help` and thus could only be addressed like that here.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java (hunk 5 above)
Comment (jjohannes) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java]: That does not work as. NameValidator is not on the classpath here (`subprojects/internal-testing`). Otherwise I would have used `NameValidator.FORBIDDEN_CHARACTERS` directly here.

**Comment by jjohannes** [comment]
@bmuschko 
> We should also mention this in the release notes as potentially breaking change.

Is that the right place to put it? It does not break anything yet as we only deprecate stuff, right? Or do we always put deprecations into that section usually?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java (hunk 9 above)
Comment (jjohannes) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java]: Yes. It did not match constructor invocations in stack traces (`<init>`). Because `<` and `>` where not matched.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java (hunk 3 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: I was starting with that but it felt pointless as it was only repeating the implementation. But I think I should add unit test coverage for the domain object classes that are affected by this, as also requested by @adammurdoch. These should then test the full character set.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java (hunk 11 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java]: Right. I will rethink this.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java (hunk 12 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java]: Yes that's better. Thanks for the suggestion.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java (hunk 11 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java]: Introduced `AbstractValidatingNamedDomainObjectContainer` and used that only for:
- Configurations
- Flavors
- Artifact Types

**Comment by jjohannes** [comment]
Added unit test coverage: https://github.com/gradle/gradle/blob/befd86505fd23d51f45f068608606c986f4ae7f2/subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java (hunk 3 above)
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: If we can test the same functionality with unit tests rather than integration tests that would be even better. Unit tests are usually faster to execute. Nevertheless, there has to be some integration tests but it can be lighter.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 10 above)
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: Maybe have two separate test cases to make it more clear why we are doing it. For the test case that cannot be run on Windows add a `@IgnoreIf` annotation.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java (hunk 5 above)
Comment (bmuschko) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/encoding/Identifier.java]: OK, thanks for clarifying.

**Comment by bmuschko** [comment]
> Is that the right place to put it? It does not break anything yet as we only deprecate stuff, right? Or do we always put deprecations into that section usually?

I was thinking about the use case that changes spaces to underscores. I am not sure if some consumer expects the "old" behavior. In that case it would be a breaking change. Either way I think we should explicitly describe the change and make people aware of it.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+import static org.gradle.internal.os.OperatingSystem.WINDOWS
+import static org.gradle.internal.os.OperatingSystem.current
+
+class NameValidationIntegrationTest extends AbstractIntegrationSpec {
+
+    def "project names should not contain forbidden characters"() {
+        when:
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: This `when` reads more like a `given` in this and the other use cases. Would you mind changing it?

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    private static final char REPLACEMENT_CHARACTER = '_';
+
+    private NameValidator() { }
+
+    /**
+     * Validates that a given name string does not contain any forbidden characters.
+     */
+    public static void validate(String name) {
+        if (StringUtils.containsAny(name, FORBIDDEN_CHARACTERS)) {
+            DeprecationLogger.nagUserOfDeprecated("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS) + ". This");
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: The use of `This` at the end feels kind of wonky. Maybe we should introduce a new method that doesn't require us to do so.

File: subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        then:
+        1 * loggingDeprecatedFeatureHandler.deprecatedFeatureUsed(_  as DeprecatedFeatureUsage) >> { DeprecatedFeatureUsage usage ->
+            assertForbidden(name, usage.message)
+        }
+
+        where:
+        [name, objectType, domainObjectContainer] << [invalidNames, domainObjectContainersWithValidation].combinations().collect { [it[0], it[1][0], it[1][1]] }
+    }
+
+    private assertForbidden(name, message) {
```
Comment (bmuschko) [inline subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy]: I think we should have two distinct assertion methods here.

File: subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy (hunk 13 above)
Comment (jjohannes) [inline subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy]: Yeah... I so much wanted `expectDeprecationWarning()` to be in the `then:` section. Which does not work with the way the fixture works. So I messed this up. I restructured the test.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java (hunk 14 above)
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: Probably. I added a method.

File: subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy (hunk 15 above)
Comment (jjohannes) [inline subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy]: Do you have a good suggestion how to do this in this multi-dimensional test setup? I feel like just asserting that it is giving _some_ deprecation warning is ok for what we test here. If it allows us to keep the test definition compact.

**Comment by jjohannes** [comment]
> I was thinking about the use case that changes spaces to underscores.

Good point. I did not think about that. Maybe we should postpone that part of the PR until 5.0. I think that might be better. So that we do not have an actual breaking change.

File: subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy (hunk 15 above)
Comment (bmuschko) [inline subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy]: I guess you'd have to spell it out meaning write out every possible combination in the `where` clause. Or maybe break up the test into multiple matrixes.

File: subprojects/core/src/main/java/org/gradle/util/NameValidator.java [hunk 16]
```diff
@@ -33,9 +33,9 @@ private NameValidator() { }
      */
     public static void validate(String name) {
         if (StringUtils.containsAny(name, FORBIDDEN_CHARACTERS)) {
-            DeprecationLogger.nagUserOfDeprecated("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS) + ". This");
+            DeprecationLogger.nagUserOfDeprecatedThing("The name '" + name + "' contains at least one of the following characters: " + Arrays.toString(FORBIDDEN_CHARACTERS));
```
Comment (bmuschko) [inline subprojects/core/src/main/java/org/gradle/util/NameValidator.java]: :o: I think we should find a better word than "thing". How about "GradleDomainObject"? 

...I just saw that we already have been using "thing" before. ;-)

File: subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java [hunk 17]
```diff
@@ -211,6 +211,12 @@ public static void nagUserOfDeprecatedBehaviour(String behaviour) {
         }
     }
 
+    public static void nagUserOfDeprecatedThing(String thing) {
+        if (isEnabled()) {
+            nagUserOfDeprecated(String.format("%s. This", thing));
```
Comment (bmuschko) [inline subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java]: :o: Maybe here use "This domain object".

**Comment by bmuschko** [comment]
> Good point. I did not think about that. Maybe we should postpone that part of the PR until 5.0. I think that might be better. So that we do not have an actual breaking change.

That might be a good idea. Maybe create a new issue for it and defer.

File: subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy (hunk 15 above)
Comment (jjohannes) [reply subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy]: I ended up inspecting the name inside `assertForbidden` and only check one message then. Everything else blew up the test code. Still now, if the test should fail, you get a better idea of what failed.

File: subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java (hunk 17 above)
Comment (jjohannes) [reply subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java]: Well, we do not deprecate the domain object. `This domain object is deprecated in...` is not true. We only deprecate certain naming. But `This naming is deprecated...` Sounds weird to me. I prefer just `This is deprecated...`, where _this_ refers to the sentence before. I do not have a better idea and I think the complete composed message for the user sounds good.

**Comment by jjohannes** [comment]
I removed the automatic adjustment of names and postponed it to 5.0 (https://github.com/gradle/gradle-private/issues/942).

I added a section about the deprecation to the release notes.

File: subprojects/docs/src/docs/release/notes.md [hunk 18]
```diff
@@ -125,9 +125,12 @@ in the next major Gradle version (Gradle 5.0). See the User guide section on the
 
 The following are the newly deprecated items in this Gradle release. If you have concerns about a deprecation, please raise it via the [Gradle Forums](https://discuss.gradle.org).
 
-<!--
-### Example deprecation
--->
+### Deprecated usage of path separator characters in names
+
+We deprecated the usage of the following characters in domain object names such as project or task names:
```
Comment (oehme) [inline subprojects/docs/src/docs/release/notes.md]: This should explicitly call out the case in `settings.gradle` where users call `include 'foo/bar'` and tell them to use `include 'foo:bar'` instead. Otherwise this warning might be non-actionable for them. Maybe our documentation for multi-projects isn't good enough, but this comes up *a lot*.

Comment (madorb) [reply subprojects/docs/src/docs/release/notes.md]: major bummer this didn't get documented. this fundamentally broke my build and i had to do a bunch of googling and i still can't figure out why renaming my projects doesn't work anymore

```include 'path:projecta' 'path:projectb'
project.(':path:projecta').name = 'projecta'
project.(':path:projectb').name = 'projectb'```

Comment (oehme) [reply subprojects/docs/src/docs/release/notes.md]: This is unrelated to this change (or should be). The code you showed could be made clearer, but it still should work afaik.

Nothing should break, we only deprecated at this point. 

Please open a bug report with a reproducible example.

Comment (jjohannes) [reply subprojects/docs/src/docs/release/notes.md]: Regarding the original comment by @oehme about documenting this better. Use this issue for further discussion: #3008
